### PR TITLE
add auto lock class

### PR DESCRIPTION
**Pull request by muleisheng**

### CODE REVIEW

**Comment by xiaoxq** [comment]
What is the usage?

**Comment by muleisheng** [comment]
> What is the usage?

I have added logic to use this auto lock.

File: modules/common/util/auto_lock.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+void InternalUnLock(T *t);
+
+/**
+ * @brief override these two routines for other synchronization methods.
+ */
+void InternalLock(std::mutex *t) noexcept(false) {
+    if (!t)
+        throw std::errc::bad_address;
+    else
+        t->lock();
```
Comment (xiaoxq) [inline modules/common/util/auto_lock.h]: 1. Always use {} for sub-scopes.
2. Fix indents to 2 spaces.

File: modules/common/util/auto_lock.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+void InternalLock(T *t);
+template<typename T>
+void InternalUnLock(T *t);
+
+/**
+ * @brief override these two routines for other synchronization methods.
+ */
+void InternalLock(std::mutex *t) noexcept(false) {
+    if (!t)
+        throw std::errc::bad_address;
```
Comment (xiaoxq) [inline modules/common/util/auto_lock.h]: Avoid "throw" if possible. Simply CHECK.

File: modules/common/util/auto_lock.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        if (!t_)
+            throw std::errc::bad_address;
+        if (!lock_) {
+            InternalLock(t_);
+            lock_ = true;
+        }
+    }
+
+ private:
+    T *t_;
```
Comment (fengqikai1414) [inline modules/common/util/auto_lock.h]: Use `T& t_` may be better

File: modules/common/util/auto_lock.h (hunk 2 above)
Comment (muleisheng) [reply modules/common/util/auto_lock.h]: OK ， I replace throw by CHECK

File: modules/common/util/auto_lock_test.cc [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#include "gtest/gtest.h"
+
+#define MAX_COUNT       10000000
+#define EXPECTED_RESULT 20000000
+
+std::mutex g_mutex;
+int g_int = 0;
+
+void function() {
+    for (int i = 0; i < MAX_COUNT; i++) {
```
Comment (xiaoxq) [inline modules/common/util/auto_lock_test.cc]: Indents as 2.

File: modules/common/util/auto_lock.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ public:
+  explicit AutoLock(std::mutex *t) :
+    t_(t),
+    lock_(false) {
+    Lock();
+  }
+
+  ~AutoLock() {
+    UnLock();
+  }
```
Comment (xiaoxq) [inline modules/common/util/auto_lock.h]: Could you try std::lock_guard? It's pretty similar with your implementation.

Comment (muleisheng) [reply modules/common/util/auto_lock.h]: auto_lock supplies unlock by manual when we need.

Comment (xiaoxq) [reply modules/common/util/auto_lock.h]: Personally I prefer a scoped lock_guard in such case. It's pretty straight forward and easy to understand by everybody.

**Pull request closed without merging**